File: ExpenseMaster/infra/src/main/java/com/expensemaster/infra/HeaderGetter.java
package com.expensemaster.infra;

import io.opentelemetry.context.propagation.TextMapGetter;
import jakarta.servlet.http.HttpServletRequest;

import java.util.Set;

public class HeaderGetter implements TextMapGetter<Object> {
    @Override
    public Iterable<String> keys(Object carrier) {
        return Set.of("traceparent");
    }

    @Override
    public String get(Object carrier, String key) {
        return ((HttpServletRequest) carrier).getHeader(key);
    }
}

File: MailDispatcher/core/src/main/kotlin/com/maildispatcher/core/HeaderGetter.kt
package com.maildispatcher.core

import io.opentelemetry.context.propagation.TextMapGetter

class HeaderGetter : TextMapGetter<Map<String, String>> {
    override fun keys(carrier: Map<String, String>): Iterable<String> {
        return carrier.keys
    }

    override fun get(carrier: Map<String, String>?, key: String): String? {
        return carrier?.get(key)
    }
}